File: app/algorithm/AA_25-26/utils/algorithm_utils.py
from typing import List

from models.channel import Channel
from models.instance_data import InstanceData
from models.program import Program
from models.schedule import Schedule
from utils.utils import Utils


class AlgorithmUtils:

    @staticmethod
    def get_best_fit(scheduled_programs: List[Schedule], instance_data: InstanceData, schedule_time: int,
                     valid_channel_indexes: List[int]) ->  tuple[Channel, Program, int]:

        # returns best channel to pick at the time and what score will it provide if we switch to it

        max_score = 0
        best_channel = None
        best_program = None

        for channel_index in valid_channel_indexes:
            channel = instance_data.channels[channel_index]
            program = Utils.get_channel_program_by_time(channel, schedule_time)

            if not program:
                continue

            score = 0

            score += program.score
            score += AlgorithmUtils.get_time_preference_bonus(instance_data, program, schedule_time)
            score += AlgorithmUtils.get_switch_penalty(scheduled_programs, instance_data, channel)
            score += AlgorithmUtils.get_delay_penalty(scheduled_programs, instance_data, program, schedule_time)
            score += AlgorithmUtils.get_early_termination_penalty(scheduled_programs, instance_data, program, schedule_time)

            if score > max_score:
                max_score = score
                best_channel = channel
                best_program = program

        return best_channel, best_program, max_score

    @staticmethod
    def get_time_preference_bonus(instance_data: InstanceData, program: Program, schedule_time: int):
        """
        Calculate time preference bonus for a program.
        CRITICAL: The bonus only applies if the program overlaps with the preferred time window
        for AT LEAST the minimum duration D minutes.

        Per PDF specification: "In order to collect the bonus, the program must fall within
        the preferred interval with at least the minimum scheduled duration D."
        """
        score = 0
        for preference in instance_data.time_preferences:
            if program.genre == preference.preferred_genre:
                # Calculate overlap between program and preference interval
                overlap_start = max(program.start, preference.start)
                overlap_end = min(program.end, preference.end)
                overlap_duration = overlap_end - overlap_start

                # Only award bonus if overlap is at least min_duration
                if overlap_duration >= instance_data.min_duration:
                    score += preference.bonus

        return score

    @staticmethod
    def get_switch_penalty(scheduled_programs: List[Schedule], instance_data: InstanceData, channel: Channel):
        penalty = 0
        if not scheduled_programs:
            return penalty

        last_schedule = scheduled_programs[-1]
        if last_schedule.channel_id != channel.channel_id:
            penalty -= instance_data.switch_penalty

        return penalty

    @staticmethod
    def get_delay_penalty(scheduled_programs: List[Schedule], instance_data: InstanceData, program: Program,
                          schedule_time: int):
        """
        Penalty for switching to a program after its scheduled start time.
        Since we now always schedule programs at their original times, this should not apply.
        """
        penalty = 0
        # No delay penalty - we always schedule programs at their original start time
        return penalty

    @staticmethod
    def get_early_termination_penalty(scheduled_programs: List[Schedule], instance_data: InstanceData, program: Program,
                                      schedule_time: int):
        """
        Penalty for terminating the previous program before its scheduled end time.
        This occurs when we switch to a new program while the previous one is still running.
        Since we now prevent overlaps, this checks if switching to the new program would
        cut off the previous one early.
        """
        penalty = 0
        if not scheduled_programs:
            return penalty

        last_schedule = scheduled_programs[-1]
        
        # If the new program starts before the previous program's natural end
        # (and we're not continuing the same program), we're cutting it short
        if last_schedule.unique_program_id != program.unique_id and program.start < last_schedule.end:
            penalty -= instance_data.termination_penalty

        return penalty

File: app/algorithm/AA_25-26/SchedulingAPI/Utils/AlgorithmUtils.cs
using SchedulingAPI.Models;

namespace SchedulingAPI.Utils
{
    public static class AlgorithmUtils
    {
        public static (Channel? bestChannel, TVProgram? bestProgram, int maxScore) GetBestFit(
            List<Schedule> scheduledPrograms, 
            InstanceData instanceData, 
            int scheduleTime,
            List<int> validChannelIndexes)
        {
            int maxScore = 0;
            Channel? bestChannel = null;
            TVProgram? bestProgram = null;

            foreach (var channelIndex in validChannelIndexes)
            {
                var channel = instanceData.Channels[channelIndex];
                var program = Utils.GetChannelProgramByTime(channel, scheduleTime);

                if (program == null) continue;

                int score = program.Score;
                score += GetTimePreferenceBonus(instanceData, program, scheduleTime);
                score += GetSwitchPenalty(scheduledPrograms, instanceData, channel);
                score += GetDelayPenalty(scheduledPrograms, instanceData, program, scheduleTime);
                score += GetEarlyTerminationPenalty(scheduledPrograms, instanceData, program, scheduleTime);

                if (score > maxScore)
                {
                    maxScore = score;
                    bestChannel = channel;
                    bestProgram = program;
                }
            }

            return (bestChannel, bestProgram, maxScore);
        }

        public static int GetTimePreferenceBonus(InstanceData instanceData, TVProgram program, int scheduleTime)
        {
            int score = 0;
            foreach (var preference in instanceData.TimePreferences)
            {
                if (program.Genre == preference.PreferredGenre)
                {
                    if (program.Start < preference.End && program.End > preference.Start)
                    {
                        score += preference.Bonus;
                    }
                }
            }
            return score;
        }

        public static int GetSwitchPenalty(List<Schedule> scheduledPrograms, InstanceData instanceData, Channel channel)
        {
            int penalty = 0;
            if (scheduledPrograms.Count == 0) return penalty;

            var lastSchedule = scheduledPrograms[^1];
            if (lastSchedule.ChannelId != channel.ChannelId)
            {
                penalty -= instanceData.SwitchPenalty;
            }

            return penalty;
        }

        public static int GetDelayPenalty(List<Schedule> scheduledPrograms, InstanceData instanceData, TVProgram program, int scheduleTime)
        {
            // No delay penalty - we always schedule programs at their original start time
            return 0;
        }

        public static int GetEarlyTerminationPenalty(List<Schedule> scheduledPrograms, InstanceData instanceData, TVProgram program, int scheduleTime)
        {
            int penalty = 0;
            if (scheduledPrograms.Count == 0) return penalty;

            var lastSchedule = scheduledPrograms[^1];

            if (lastSchedule.UniqueProgramId != program.UniqueId && program.Start < lastSchedule.End)
            {
                penalty -= instanceData.TerminationPenalty;
            }

            return penalty;
        }
    }
}

